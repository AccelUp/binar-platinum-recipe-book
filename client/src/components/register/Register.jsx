import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import clasess from "../login/Login.module.css";

const Register = () => {
  const navigate = useNavigate();

  const [formData, setFormData] = useState({
    username: "",
    password: "",
    confirmPassword: "",
  });
  const [errors, setErrors] = useState({});

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setFormData({ ...formData, [name]: value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const newErrors = {};

    if (formData.username === "") {
      newErrors.username = "username is required";
    }

    if (formData.password === "") {
      newErrors.password = "Password is required";
    }

    if (formData.confirmPassword === "") {
      newErrors.confirmPassword = "Confirm Password is required";
    }

    if (formData.password !== formData.confirmPassword) {
      newErrors.confirmPassword = "Passwords do not match";
    }

    setErrors(newErrors);

    if (Object.keys(newErrors).length === 0) {
      try {
        const response = await fetch(
          "http://localhost:8000/api/v1/recipebook/user/register",
          {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(formData),
          }
        );

        if (response.ok) {
          navigate("/login");
        } else {
          console.error("Failed to create user.");
        }
      } catch (error) {
        console.error("Failed to fetch user data.", error);
      }
    }
  };

  return (
    <>
      <div className={`${clasess.Login} relative pt-20`}>
        <div className="flex flex-col items-center justify-center px-6 py-10 mx-auto md:h-screen lg:py-0">
          <div className="w-full bg-black rounded-md shadow md:mt-0 sm:max-w-md xl:p-0">
            <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
              <h1 className="text-center text-primary font-bold leading-tight tracking-tight md:text-2xl">
                Sign up
              </h1>
              <form className="space-y-4" onSubmit={handleSubmit}>
                <div className="form-input">
                  <label className="block mb-2 font-medium text-primary">
                    Your username
                  </label>
                  <input
                    type="username"
                    name="username"
                    value={formData.username}
                    onChange={handleInputChange}
                    className={`bg-white border border-gray-300 text-gray-900 sm:text-sm rounded-md focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 ${
                      errors.username ? "border-red-500" : ""
                    }`}
                    placeholder="username"
                    required=""
                  />
                  {errors.username && (
                    <p className="text-red-500 text-xs">{errors.username}</p>
                  )}
                </div>
                <div className="form-input">
                  <label className="block mb-2 font-medium text-primary">
                    Password
                  </label>
                  <input
                    type="password"
                    name="password"
                    value={formData.password}
                    onChange={handleInputChange}
                    placeholder="••••••••"
                    className={`bg-white border border-gray-300 text-gray-900 sm:text-sm rounded-md focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5${
                      errors.password ? " border-red-500" : ""
                    }`}
                    required=""
                  />
                  {errors.password && (
                    <p className="text-red-500 text-xs">{errors.password}</p>
                  )}
                </div>
                <div className="form-input pb-5">
                  <label className="block mb-2 font-medium text-primary">
                    Confirm password
                  </label>
                  <input
                    type="password"
                    name="confirmPassword"
                    value={formData.confirmPassword}
                    onChange={handleInputChange}
                    placeholder="••••••••"
                    className={`bg-white border border-gray-300 text-gray-900 sm:text-sm rounded-md focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 ${
                      errors.confirmPassword ? "border-red-500" : ""
                    }`}
                    required=""
                  />
                  {errors.confirmPassword && (
                    <p className="text-red-500 text-xs">
                      {errors.confirmPassword}
                    </p>
                  )}
                </div>

                <button
                  type="submit"
                  className="w-full text-white bg-primary hover:bg-black hover:text-primary focus:ring-4 focus:outline-none focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center dark:bg-primary-600 dark:hover-bg-primary-700 dark:focus:ring-primary-800"
                >
                  Create an account
                </button>
                <p className="text-sm font-light text-white pb-4">
                  Already have an account?
                  <a
                    href="/login"
                    className="font-medium text-primary-600 text-primary"
                  >
                    Login here
                  </a>
                </p>
              </form>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default Register;
