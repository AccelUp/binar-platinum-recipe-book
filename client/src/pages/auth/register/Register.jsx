import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { register } from "../../../actions/authAction";
import classes from "./Register.module.css";

const Register = () => {
  const navigate = useNavigate();
  const [errorText, setErrorText] = useState("");
  const [formData, setFormData] = useState({
    username: "",
    password: "",
    confirmPassword: "",
  });

  const dispatch = useDispatch();

  const navigateToLogin = () => {
    navigate("/login");
  };

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setFormData({ ...formData, [name]: value });
  };

  const fetchRegister = async (e) => {
    e.preventDefault();

    if (
      formData.username === "" ||
      formData.password === "" ||
      formData.confirmPassword === ""
    ) {
      setErrorText("Username, password, and confirmPassword are required");
    } else if (formData.password !== formData.confirmPassword) {
      setErrorText("Passwords do not match");
    } else {
      try {
        const response = await dispatch(register(formData, navigate));

        if (response.error) {
          setErrorText("Failed to register. Please check your credentials");
        }
      } catch (error) {
        console.error("Failed to register", error);
        setErrorText("An error occurred while registering.");
      }
    }
  };

  return (
    <div className={`${classes.Register} relative pt-20`}>
      <div className="flex flex-col items-center justify-center px-6 py-10 mx-auto md:h-screen lg:py-0">
        <div className="w-full bg-black rounded-md shadow md:mt-0 sm:max-w-md xl:p-0">
          <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
            <h1 className="text-center text-primary font-bold leading-tight tracking-tight md:text-2xl">
              Sign up
            </h1>
            <form className="space-y-4" onSubmit={fetchRegister}>
              <div className="mb-4">
                <label
                  className="block font-semibold text-primary mb-2"
                  htmlFor="username"
                >
                  Username
                </label>
                <input
                  className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                    errorText ? "border-red-500" : ""
                  }`}
                  id="username"
                  name="username"
                  value={formData.username}
                  type="text"
                  onChange={handleInputChange}
                  placeholder="Enter your username"
                />
              </div>
              <div className="mb-4">
                <label
                  className="block font-semibold text-primary mb-2"
                  htmlFor="password"
                >
                  Password
                </label>
                <input
                  className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                    errorText ? "border-red-500" : ""
                  }`}
                  id="password"
                  name="password"
                  type="password"
                  value={formData.password}
                  onChange={handleInputChange}
                  placeholder="Enter your password"
                />
              </div>
              <div className="mb-4">
                <label
                  className="block font-semibold text-primary mb-2"
                  htmlFor="confirmPassword"
                >
                  Confirm Password
                </label>
                <input
                  className={`border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline ${
                    errorText ? "border-red-500" : ""
                  }`}
                  id="confirmPassword"
                  name="confirmPassword"
                  type="password"
                  value={formData.confirmPassword}
                  onChange={handleInputChange}
                  placeholder="Confirm your password"
                />
              </div>
              {errorText && (
                <div className="text-red-500 mb-4 p-2 bg-red-100 border border-red-400 rounded">
                  {errorText}
                </div>
              )}
              <button
                className="bg-primary hover:bg-black hover:text-primary text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                type="submit"
              >
                Register
              </button>
            </form>
            <h2 className="mb-2 text-sm pt-2">
              Already have an account?{" "}
              <span
                onClick={navigateToLogin}
                className="text-primary cursor-pointer"
              >
                Login
              </span>
            </h2>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Register;
