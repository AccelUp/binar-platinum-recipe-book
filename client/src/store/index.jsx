import { createStore, applyMiddleware } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import thunk from "redux-thunk";
import rootReducer from "../reducers";
import { composeWithDevTools } from "redux-devtools-extension";

// Konfigurasi Redux Persist
const persistConfig = {
  key: "root", // Key untuk penyimpanan
  storage, // Jenis penyimpanan (local storage)
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Buat store dengan middleware dan reducer yang telah di-persist
const store = createStore(
  persistedReducer,
  composeWithDevTools(applyMiddleware(thunk))
);
const persistor = persistStore(store);

export { store, persistor };
