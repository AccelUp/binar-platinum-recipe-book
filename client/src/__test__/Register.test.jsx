import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../store"; // Import your Redux store configuration
import renderer from "react-test-renderer";
import Register from "../pages/auth/register/Register";
import { applyMiddleware, createStore } from "redux"; // Import createStore from Redux
import thunk from "redux-thunk";
import rootReducer from "../reducers";

describe("Register Component", () => {
  it("should render the Register component correctly", () => {
    const reduxStore = createStore(rootReducer, applyMiddleware(thunk));

    const component = renderer.create(
      <Provider store={reduxStore}>
        <MemoryRouter>
          <Register />
        </MemoryRouter>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
