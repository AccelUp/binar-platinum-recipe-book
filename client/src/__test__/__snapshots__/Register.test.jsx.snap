// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Register Component > should render the Register component correctly 1`] = `
<div
  className="_Register_424735 relative pt-20"
>
  <div
    className="flex flex-col items-center justify-center px-6 py-10 mx-auto md:h-screen lg:py-0"
  >
    <div
      className="w-full bg-black rounded-md shadow md:mt-0 sm:max-w-md xl:p-0"
    >
      <div
        className="p-6 space-y-4 md:space-y-6 sm:p-8"
      >
        <h1
          className="text-center text-primary font-bold leading-tight tracking-tight md:text-2xl"
        >
          Sign up
        </h1>
        <form
          className="space-y-4"
          onSubmit={[Function]}
        >
          <div
            className="mb-4"
          >
            <label
              className="block font-semibold text-primary mb-2"
              htmlFor="username"
            >
              Username
            </label>
            <input
              className="border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
              id="username"
              name="username"
              onChange={[Function]}
              placeholder="Enter your username"
              type="text"
              value=""
            />
          </div>
          <div
            className="mb-4"
          >
            <label
              className="block font-semibold text-primary mb-2"
              htmlFor="password"
            >
              Password
            </label>
            <input
              className="border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
              id="password"
              name="password"
              onChange={[Function]}
              placeholder="Enter your password"
              type="password"
              value=""
            />
          </div>
          <div
            className="mb-4"
          >
            <label
              className="block font-semibold text-primary mb-2"
              htmlFor="confirmPassword"
            >
              Confirm Password
            </label>
            <input
              className="border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline "
              id="confirmPassword"
              name="confirmPassword"
              onChange={[Function]}
              placeholder="Confirm your password"
              type="password"
              value=""
            />
          </div>
          <button
            className="bg-primary hover:bg-black hover:text-primary text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
            type="submit"
          >
            Register
          </button>
        </form>
        <h2
          className="mb-2 text-sm pt-2"
        >
          Already have an account?
           
          <span
            className="text-primary cursor-pointer"
            onClick={[Function]}
          >
            Login
          </span>
        </h2>
      </div>
    </div>
  </div>
</div>
`;
